astropy/cosmology/io/table.py

```diff
@@ -119,46 +126,53 @@ def from_table(table, index=None, *, move_to_meta=False, cosmology=None):
         elif len(indices) == 1:  # only happens w/ np.where
             index = indices[0]
         elif len(indices) == 0:  # matches from loc_indices
-            raise KeyError(f"No matches found for key {indices}")
+            msg = f"No matches found for key {indices}"
```

Comment:
This is another change that is not related to the PR and, as far as I can see, just adds line and complexity (however slight) for no obvious benefit.

Reply:
Agreed it's not so related to the PR, but it does add benefit.
Python includes the raise in the traceback. By using a string literal, the error message is duplicated in the traceback, which is less readable. I tripped this in a test and wanted to improve the user experience.
It would be better if Python had an option to not include the string literal in the traceback, but short of a change to Python, this is the way to make nicer tracebacks.

Reply:
I'm not sure; what is wrong with the following traceback?
```
n [4]: u.Quantity('bla')
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
File /usr/lib/python3/dist-packages/astropy/units/quantity.py:490, in Quantity.__new__(cls, value, unit, dtype, copy, order, subok, ndmin)
    489 try:
--> 490     value = float(v.group())
    492 except Exception:

AttributeError: 'NoneType' object has no attribute 'group'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
Cell In [4], line 1
----> 1 u.Quantity('bla')

File /usr/lib/python3/dist-packages/astropy/units/quantity.py:493, in Quantity.__new__(cls, value, unit, dtype, copy, order, subok, ndmin)
    490     value = float(v.group())
    492 except Exception:
--> 493     raise TypeError(
    494         f'Cannot parse "{value}" as a {cls.__name__}. It does not '
    495         "start with a number."
    496     )
    498 unit_string = v.string[v.end() :].strip()
    499 if unit_string:

TypeError: Cannot parse "bla" as a Quantity. It does not start with a number.

```

Reply:
With the change this should look like 

```
>>> import astropy.units as u
>>> u.Quantity("bla")
Traceback (most recent call last):
  File "/Users/nmrs/local/astropy/astropy/units/quantity.py", line 479, in __new__
    value = float(v.group())
AttributeError: 'NoneType' object has no attribute 'group'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/nmrs/local/astropy/astropy/units/quantity.py", line 486, in __new__
    raise TypeError(msg)
TypeError: Cannot parse "bla" as a Quantity. It does not start with a number.
```,

getting rid of the duplication of the Exception message. We see the actual  `TypeError`, there's not much point showing it's generation during the traceback.

Reply:
Python has been careful to not show extraneous information:

```
Traceback (most recent call last):
  File "/Users/nmrs/local/astropy/astropy/units/quantity.py", line 479, in __new__
    value = float(v.group())
AttributeError: 'NoneType' object has no attribute 'group'
```

doesn't show how AttributeError is produced. Unfortunately they use C stuff to get rid of even the `raise AttributeError(msg)` step and I don't think we can achieve this degree of concision.

Reply:
Since the error message give context, there is actually very little difference. If I change `Quantity` to use `msg`, I get:
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
File ~/data/mhvk/packages/astropy/astropy/units/quantity.py:479, in Quantity.__new__(cls, value, unit, dtype, copy, order, subok, ndmin)
    478 try:
--> 479     value = float(v.group())
    481 except Exception:

AttributeError: 'NoneType' object has no attribute 'group'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
Cell In[2], line 1
----> 1 u.Quantity('bla')

File ~/data/mhvk/packages/astropy/astropy/units/quantity.py:486, in Quantity.__new__(cls, value, unit, dtype, copy, order, subok, ndmin)
    481 except Exception:
    482     msg = (
    483         f'Cannot parse "{value}" as a {cls.__name__}. It does not '
    484         "start with a number."
    485     )
--> 486     raise TypeError(msg)
    488 unit_string = v.string[v.end() :].strip()
    489 if unit_string:

TypeError: Cannot parse "bla" as a Quantity. It does not start with a number.
```

Reply:
Your view is different than mine! I made mine from the command line and it does not include the above context.

Reply:
I'll roll this back for now. We can have discussion on https://beta.ruff.rs/docs/rules/#flake8-errmsg-em in a separate Issue. It has an auto fix, so if we decide to turn it on, it's very easily maintained.

Reply:
I've reverted here.